A polynya test case

einola opened the pull request:
# A polynya test case

Fixes #375 

### Task List
- [x] Defined the tests that specify a complete and functioning change (*It may help to create a [design specification & test specification](../../../wiki/Specification-Template)*)
- [x] Implemented the source code change that satisfies the tests
- [x] Documented the feature by providing worked example
- [x] Updated the README or other documentation
- [x] Completed the pre-Request checklist below

---
# Change Description

This PR provides a polynya test case (as per issue #375). For this to work, we also had to update the handling of open boundaries. The ice now flows freely out of any open boundaries. Regarding code changes, the most important changes are to the boundary handling, while the polynya test-case setup is valuable as a test for the new boundaries. This PR also includes minor fixes and upgrades to the thermodynamics.

---
# Test Description

The configuration in ``config_polynya.cfg`` returns results visually similar to those already published (see issue #375).

---
# Documentation Impact

None.

---
# Other Details

None

---
### Pre-Request Checklist

- [x] The requirements of this pull request are fully captured in an issue or design specification and are linked and summarised in the description of this PR
- [x] No new warnings are generated
- [x] The documentation has been updated (or an issue has been created to track the corresponding change)
- [x] Methods and Tests are commented such that they can be understood without having to obtain additional context
- [x] This PR/Issue is labelled as a bug/feature/enhancement/breaking change
- [x] File dates have been updated to reflect modification date
- [x] This change conforms to the conventions described in the README

einola committed:
Polynya test case

Initial commit of a python script to create the setup and a config file
to run the test.

TODO:
 - Change the wind direction
 - Test BBM
 - Hourly output doesn't work

einola committed:
Minor updates to the config file

Most importantly, the wind is blowing at an angle now.

einola committed:
Create init file with CMake

einola committed:
Merge remote-tracking branch 'origin/develop' into issue375_polynya_test

einola committed:
Merge branch 'develop' into issue375_polynya_test

# Conflicts:
#	run/CMakeLists.txt

einola committed:
Add open (Neumann zero) boundary conditions

Adapts the dirichletFromEdge function to be neumannFromEdge and adds two
neumannZero functions - one for DGVectors and one for CGVectors. Both
are inspired by the existing dirichletZero function. The neumannZero
function must be called after advection - which is a bit worse, but
moving it into the DGTransport class is compicated.

einola committed:
Revert "Add open (Neumann zero) boundary conditions"

This reverts commit b99de51a24ca57c1d27f4328f34d6c0aeb7f1c7c.

einola committed:
Merge branch '793landmask_cg' into issue375_polynya_test

winzerle committed:
New boundary handling.

Dirichlet: all done using the cglandmask field
Neumann: can only appear on the outer edges of the mesh, if the last elements are ice elements.

winzerle committed:
- Removed old Dirichlet-Info from the mesh.
- fixed advection and mesh test cases

einola committed:
Merge branch 'develop' into issue375_polynya_test

# Conflicts:
#	dynamics/src/CGDynamicsKernel.cpp
#	dynamics/src/include/BrittleCGDynamicsKernel.hpp
#	dynamics/src/include/VPCGDynamicsKernel.hpp

einola committed:
Merge branch 'develop' into issue375_polynya_test

einola committed:
Remove VTK output from VPCGDynamicsKernel

It was there for debugging purposes only.

einola committed:
Add u and v wind direction to ConfiguredAtmosphere

This makes comparison with the published works easier, as they use
something similar to ConfiguredAtmosphere, not FluxConfiguredAtmosphere.

einola committed:
Initialise damage to one

And don't warn about zeros any more. The default rheology is still mEVP
(as this is directly comparable to the published work).

einola committed:
Fix open-water flux in FluxConfiguredAtmosphere

The open water heat flux is reset by the thermodynamics, that the
(slab) ocean doesn't super cool. This commit sets it back to the initial
value on every FluxConfiguredAtmosphere::update call.

einola committed:
Update config_polynya.cfg

Run for the full five days (as previous publications did), use Winton
thermodynamics (again) and tweak the ice-atmosphere and open-water heat
fluxes.

einola committed:
Minor tweaks and fixes

Added a lot of const-s, changed one equation to be more readable (for me
at least), modified the limits to include fixing negative concentration
and thickness values, and removed redundant qualifiers.

einola committed:
Merge branch 'develop' into issue375_polynya_test

einola committed:
Clang-format fixes

The test failed, so I ran clang-format on the files.

einola committed:
Increase the domain size

This is the same now as in Bjornsson et al.

einola committed:
Update PrognosticData_test.cpp

ConfiguredAtmosphere doesn't accept wind_speed as an option anymore,
only wind_u and wind_v, so I set them to 3 and 4 to get a wind speed of
5. Also removed an unused header file.

einola committed:
Remove commented-out code

It's not needed any more! Also some clang-format changes.

einola committed:
Merge remote-tracking branch 'origin/develop' into issue375_polynya_test

einola committed:
Minor fixes to get the model to compile and polynya simulation to start

But it still crashes with unstabilities on the boundary after a few time
steps.

einola committed:
Merge branch 'develop' into issue375_polynya_test

timspainNERSC commented on physics/src/IceGrowth.cpp:
Since the bulk of this PR covers dynamics and boundary conditions, these changes should be a separate PR, especially as they don't change the numerical outcome of the `IceGrowth` calculations (I think).

timspainNERSC replied:
And if they do change the numerical output, that value should be tested in the `IceGrowth` test.

timspainNERSC commented on core/test/PrognosticData_test.cpp:
These changes do not directly impact the polynya test, and should be a separate PR

timspainNERSC commented on physics/src/modules/AtmosphereBoundaryModule/FluxConfiguredAtmosphere.cpp:
```diff
@@ -108,4 +108,11 @@ void FluxConfiguredAtmosphere::setData(const ModelState::DataMap& dm)
     vwind = v0;
 }
 
+void FluxConfiguredAtmosphere::update(const TimestepTime& tst)
+{
+    /* The open water heat flux is reset by the thermodynamics, so that the (slab) ocean doesn't
+     * super cool. Therefore, we have to reset it here on every update */
+    qow = qow0;
+}
+
```
Is this required for the polynya test? It should go in the PR with the `IceGrowth` changes if it is not required for the polynya test to pass.

einola replied:
The configuration can use either ``ConfiguredAtmosphere`` or ``FluxConfiguredAtmosphere``. So far, I've used ``FluxConfiguredAtmosphere`` because it's simpler, but in the literature, people have used the equivalence of ``ConfiguredAtmosphere``, so I've been switching between the two.

The bottom line is that both need these minor fixes for the configuration to work, so I think it's okay to include both in the PR, even though only one will be used in the "official" setup.

timspainNERSC commented on physics/src/modules/AtmosphereBoundaryModule/ConfiguredAtmosphere.cpp:
If these changes aren't required by the polynya test, they should go in a separate PR. (I think in this case they might be required to force the polynya dynamics correctly).

einola replied:
Yes, these are not relevant or necessary, so I'm reverting them.

timspainNERSC commented on physics/src/modules/AtmosphereBoundaryModule/include/ConfiguredAtmosphere.hpp:
See comment on the corresponding `.cpp` file.

timspainNERSC commented on physics/src/modules/AtmosphereBoundaryModule/include/FluxConfiguredAtmosphere.hpp:
```diff
@@ -42,8 +42,8 @@ class FluxConfiguredAtmosphere : public IAtmosphereBoundary,
 protected:
-    //! Performs the implementation specific updates. Does nothing.
-    void update(const TimestepTime&) override { }
+    //! Performs the implementation specific updates.
+    void update(const TimestepTime&) override;
```
See comment on the corresponding .cpp file.

timspainNERSC reviewed (changes requested):
Some irrelevant changes coming along for the ride.

einola commented:
Merging develop into this branch borked the whole thing - we now have instability on the boundary again, and I don't understand why.

einola committed:
Revert "Minor tweaks and fixes"

This reverts commit 768c4d4c

einola committed:
Merge remote-tracking branch 'origin/develop' into issue375_polynya_test

einola committed:
Merge remote-tracking branch 'origin/develop' into issue375_polynya_test

# Conflicts:
#	dynamics/test/AdvectionPeriodicBC_test.cpp
#	dynamics/test/Advection_test.cpp
#	dynamics/test/ParametricMesh_test.cpp

einola committed:
Change back dates on several files

With the latest merges, the date is the only difference between this
branch and develop ... and that doesn't make sense.

einola committed:
Fix the date fixing

Got the year wrong on one of the files

einola committed:
Merge remote-tracking branch 'origin/develop' into issue375_polynya_test

einola committed:
Merge remote-tracking branch 'origin/develop' into issue375_polynya_test

einola committed:
Fix polynya initialisation

The call to initMaker in make_init_polynya.py was still using nLayers,
and not isWinton.

einola committed:
Make an integration test out of the polynya case

I reduce the resolution, increase the time step, and run it for five
days. This results in a steady state. I then simply compare the max,
min, and mean of hice, cice, u, and v with "known good values".

einola committed:
Additional python path for the CI

The setup is different over there.

einola committed:
Fix wrong python command on macOS

Copy/paste error and python vs python3

einola committed:
Reduce accuracy of polynya integration test

This is much more sensitive than I expected. So, I'm only using the DG0
component and not checking the velocities. Running it on macOS vs Linux
gives fairly different maximum values for velocities and higher order DG
fields.

einola committed:
Reduce the accuracy even further

But it's still useful

einola commented:
Now contains only directly polynya-related changes (i.e. config and init files, and an integrtion test)

joewallwork commented on run/make_init_polynya.py:
```diff
@@ -0,0 +1,50 @@
+nsecond = int(x / res)
+nLayers = 3
+
+fname = f"init_polynya.nc"
```
Redundant f-string.
```suggestion
fname = "init_polynya.nc"
```

einola replied:
Done

joewallwork commented on run/make_init_polynya.py:
```diff
@@ -0,0 +1,50 @@
+res = 2
+
+
+nfirst = int(y / res)
+nsecond = int(x / res)
```
How about integer division specifically, to avoid the type conversion?
```suggestion
nfirst = y // res/
nsecond = x // res
```

einola replied:
Done

joewallwork commented on test/PolynyaIntegration_test.py:
```diff
@@ -0,0 +1,185 @@
+import numpy as np
+
+
+class Polynya(unittest.TestCase):
```
It'd be useful to have a docstring that gives an overview of what the test case is. It doesn't necessarily need to go into details, but could provide a link to where to get more info. e.g., the paper referenced elsewhere.

einola replied:
Done

joewallwork commented on test/PolynyaIntegration_test.py:
```diff
@@ -0,0 +1,185 @@
+        res = 4
+
+
+        nfirst = int(y / res)
+        nsecond = int(x / res)
```
```suggestion
        nfirst = y // res
        nsecond = x // res
```

einola replied:
Done

joewallwork commented on test/PolynyaIntegration_test.py:
```diff
@@ -0,0 +1,185 @@
+        nfirst = int(y / res)
+        nsecond = int(x / res)
+
+        fname = f"init_polynya.nc"
```
```suggestion
        fname = "init_polynya.nc"
```

einola replied:
Done

joewallwork commented on test/PolynyaIntegration_test.py:
```diff
@@ -0,0 +1,185 @@
+
+        fname = f"init_polynya.nc"
+
+        # The model expects everything in metres
+        initializer = initMaker(fname, nfirst, nsecond, res*1e3, isWinton=True, checkZeros=False)
+
+        # Ice everywhere and all boundaries closed, except the x = 100 km end
+        initializer.mask[:, :] = 1.
+        initializer.mask[0, :] = 0.
+        initializer.mask[-1, :] = 0.
+        initializer.mask[:, 0] = 0.
+        #initializer.mask[:, -1] = 0. ## right
+
+        # Uniform concentration of 90%
+        initializer.cice[:, :] = 0.9
+
+        # Uniform thickness of 20 cm
+        initializer.hice[:, :] = 0.2
+
+        # Undamaged ice
+        initializer.damage[:, :] = 1.
+
+        # Ice and ocean temperature and salinity at the freezing point
+        ice_salinity = 5  # should match Ice::s in constants.hpp
+        mu: float = -0.055  # should match Water::mu in constants.hpp
+        ocean_temperature = -1.54
+        ocean_salinity = ocean_temperature / mu
+
+        initializer.sss[:, :] = ocean_salinity
+        initializer.sst[:, :] = ocean_temperature
+        initializer.tsurf[:, :] = ice_salinity * mu
+        initializer.tbott = initializer.tsurf
+        initializer.tintr = initializer.tsurf
```
This code seems to be duplicated in `make_init_polynya.py`. It'd be nice to avoid that, if possible.

einola replied:
Yes, but I'm not sure how to do that sensibly. The test in ``test/PolynyaIntegration_test.py``should not be touched, while the script in ``run/make_init_polynya.py`` should be an example of how to run one of the cases in the Bjornsson et al paper. So, I don't see how one could reuse code from the other.

joewallwork replied:
That's true, I suppose it'd need to be a more significant refactor than would be worth doing here.

joewallwork reviewed (changes requested):
Thanks, this mostly looks good but I have some minor suggestions and a slightly bigger one to avoid duplicated code.

Note that the changes I'm making in #889 will rework the approach taken for scripts in the `run` directory, but I'm happy to account for this new addition later in my changes.

einola committed:
Remove redundant f-string from make_init_polynya.py

As suggested by Joe on github.

Co-authored-by: Joe Wallwork <22053413+jwallwork23@users.noreply.github.com>

einola committed:
Use integer division, instead of type conversion

As suggested by Joe on github

einola committed:
Use the class variable for init_file in PolynyaIntegration_test.py

That's cleaner than using a fixed string.

einola committed:
Document the PolynyaIntegration_test.py with a docstring

As suggested by Joe on github

einola commented:
Funnily enough, the test now fails because of changes from PR #884. So, if this current PR had been merged first, then we wouldn't have merged #884 the way it was! I've asked Thomas to take a look!

einola commented:
The tests should pass again after #893 is merged into develop.

einola committed:
Merge remote-tracking branch 'origin/develop' into issue375_polynya_test

joewallwork reviewed (approved):
Thanks for the updates to this @einola, I'm happy with it now.

einola committed:
Merge branch 'develop' into issue375_polynya_test

einola committed:
Update PolynyaIntegration_test.py "known good values"

This test seems almost too sensitive to small changes in the model.
Changing the advection scheme caused minor changes in the results; they
aren't physically important, but the mean thickness and concentration
changed a bit. I don't think this invalidates the test. The mean
concentration also changes by 0.001 between platforms.

einola commented:
@timspainNERSC: Can you have a look at this one? I can't merge without your approval, because you asked for changes a while ago.

timspainNERSC commented on run/make_init_polynya.py:
```diff
@@ -0,0 +1,49 @@
+
+nfirst = y // res
+nsecond = x // res
+nLayers = 3
```
Don't need this any more!

timspainNERSC reviewed (approved):
Looks good, but the `nLayers` variable in the python script that creates the initial file is no longer needed.

einola committed:
Merge branch 'develop' into issue375_polynya_test

einola committed:
Update the polynya integration test script

Several changes in develop require coding changes in the script. In
particular, there are no netCDF groups in the outputs any more, and
temperature is not explicitly initialised when using initMaker.
